File: frontend/src/modules/orders/calculations.ts
export interface PreviewInput {
  settlementBasis?: 'ORDER_AMOUNT' | 'AFTER_COUPON'
  settlementMethod?: 'DISCOUNT' | 'FIXED_DEDUCTION'
  discount?: number
  fixedDeduction?: number
  orderAmount?: number
  couponAmount?: number
  actualPaid?: number
  defaultCommission?: number
  settlementIncomeOverride?: number
  commissionOverride?: number
}

export function calculateDefaultActualPaid(
  orderAmount?: number | null,
  couponAmount?: number | null,
) {
  if (orderAmount === undefined || orderAmount === null || couponAmount === undefined || couponAmount === null) {
    return undefined
  }
  const difference = Number(orderAmount) - Number(couponAmount)
  if (!Number.isFinite(difference)) return undefined
  return Math.max(0, Math.round((difference + Number.EPSILON) * 100) / 100)
}

export function calculateOrderPreview(input: PreviewInput) {
  const orderAmount = Number(input.orderAmount ?? 0)
  const couponAmount = Number(input.couponAmount ?? 0)
  const actualPaid = Number(input.actualPaid ?? 0)
  const basisAmount = input.settlementBasis === 'AFTER_COUPON'
    ? orderAmount - couponAmount
    : orderAmount
  const defaultIncome = input.settlementMethod === 'FIXED_DEDUCTION'
    ? Math.max(0, basisAmount - Number(input.fixedDeduction ?? 0))
    : basisAmount * Number(input.discount ?? 0)
  const income = input.settlementIncomeOverride ?? defaultIncome
  const commission = input.commissionOverride ?? Number(input.defaultCommission ?? 0)
  const cost = actualPaid + commission
  return { income, commission, cost, profit: income - cost }
}

File: backend/app/modules/orders/calculations.py
from dataclasses import dataclass
from decimal import ROUND_HALF_UP, Decimal

from app.modules.partners.models import SettlementBasis

CENT = Decimal("0.01")


def money(value: Decimal) -> Decimal:
    return Decimal(value).quantize(CENT, rounding=ROUND_HALF_UP)


@dataclass(frozen=True)
class OrderAmounts:
    settlement_income: Decimal
    commission: Decimal
    cost: Decimal
    profit: Decimal


def calculate_order_amounts(
    *,
    order_amount: Decimal,
    coupon_amount: Decimal,
    actual_paid: Decimal,
    settlement_basis: SettlementBasis,
    discount: Decimal,
    commission: Decimal,
    settlement_income_override: Decimal | None = None,
) -> OrderAmounts:
    if min(order_amount, coupon_amount, actual_paid, commission) < 0:
        raise ValueError("金额不能为负数")
    if coupon_amount > order_amount:
        raise ValueError("优惠券金额不能超过订单标价")
    if discount <= 0 or discount > 1:
        raise ValueError("折扣必须大于0且不超过1")

    after_coupon = money(order_amount - coupon_amount)
    basis_amount = order_amount if settlement_basis == SettlementBasis.ORDER_AMOUNT else after_coupon
    excess_amount = max(order_amount - Decimal("210"), Decimal("0"))
    default_income = basis_amount * discount + excess_amount * (Decimal("1") - discount)
    income = (
        money(settlement_income_override)
        if settlement_income_override is not None
        else money(default_income)
    )
    final_commission = money(commission)
    cost = money(actual_paid + final_commission)
    profit = money(income - cost)
    return OrderAmounts(income, final_commission, cost, profit)